File: src/vector_tongue/geometry.py
"""Geometric drift and similarity measures for embedding vectors."""

import math
from typing import Sequence, Dict


def dot_product(a: Sequence[float], b: Sequence[float]) -> float:
    """Compute dot product between two vectors."""
    if len(a) != len(b):
        raise ValueError(f"Vector dimension mismatch: {len(a)} vs {len(b)}")
    return sum(x * y for x, y in zip(a, b))


def vector_norm(a: Sequence[float]) -> float:
    """Compute Euclidean (L2) norm of a vector."""
    return math.sqrt(sum(x * x for x in a))


def cosine_similarity(a: Sequence[float], b: Sequence[float]) -> float:
    """Compute cosine similarity between two vectors."""
    norm_a = vector_norm(a)
    norm_b = vector_norm(b)
    if norm_a == 0.0 or norm_b == 0.0:
        return 0.0
    val = dot_product(a, b) / (norm_a * norm_b)
    # Clamp to [-1.0, 1.0] to prevent numerical floating point drift
    return max(-1.0, min(1.0, val))


def cosine_distance(a: Sequence[float], b: Sequence[float]) -> float:
    """Compute cosine distance: 1 - cosine_similarity(a, b)."""
    return 1.0 - cosine_similarity(a, b)


def euclidean_distance(a: Sequence[float], b: Sequence[float]) -> float:
    """Compute Euclidean distance ||b - a||_2."""
    if len(a) != len(b):
        raise ValueError(f"Vector dimension mismatch: {len(a)} vs {len(b)}")
    return math.sqrt(sum((y - x) ** 2 for x, y in zip(a, b)))


def norm_shift(a: Sequence[float], b: Sequence[float]) -> float:
    """Compute change in vector length: ||b||_2 - ||a||_2."""
    return vector_norm(b) - vector_norm(a)


def marler_drift_v1(a: Sequence[float], b: Sequence[float]) -> float:
    """Historical Marler Drift v1 index: (1 - cos(a, b)) * ||b - a||_2."""
    return cosine_distance(a, b) * euclidean_distance(a, b)


def compute_pair_metrics(a: Sequence[float], b: Sequence[float]) -> Dict[str, float]:
    """Compute comprehensive descriptive metrics for a source/target vector pair."""
    return {
        "cosine_similarity": cosine_similarity(a, b),
        "cosine_distance": cosine_distance(a, b),
        "euclidean_distance": euclidean_distance(a, b),
        "norm_shift": norm_shift(a, b),
        "marler_drift_v1": marler_drift_v1(a, b),
    }

File: src/lib/geometry.ts
/**
 * Numerically safe geometry and historically versioned drift measures.
 */
import { dot, norm, Vector } from "./matrix";

export interface DriftComponents {
  cosine_distance: number;
  euclidean_displacement: number;
  norm_shift: number;
  marler_drift_v1: number;
}

export function asVector(value: unknown, name = "vector"): Vector {
  if (!Array.isArray(value)) {
    throw new Error(`${name} must be an array`);
  }
  if (value.length === 0) {
    throw new Error(`${name} must not be empty`);
  }
  const vec = value.map((v, i) => {
    const num = Number(v);
    if (!Number.isFinite(num)) {
      throw new Error(`${name} contains a non-finite value at index ${i}`);
    }
    return num;
  });
  return vec;
}

export function requireSameShape(a: Vector, b: Vector): void {
  if (a.length !== b.length) {
    throw new Error(`vectors must have equal shape; got ${a.length} and ${b.length}`);
  }
}

export function cosineSimilarity(a: Vector, b: Vector, epsilon = 1e-12): number {
  requireSameShape(a, b);
  const normA = norm(a);
  const normB = norm(b);
  const denominator = normA * normB;
  if (denominator <= epsilon) {
    throw new Error("cosine similarity is undefined for a near-zero vector");
  }
  const cos = dot(a, b) / denominator;
  return Math.max(-1.0, Math.min(1.0, cos));
}

export function cosineDistance(a: Vector, b: Vector): number {
  return 1.0 - cosineSimilarity(a, b);
}

export function l2Distance(a: Vector, b: Vector): number {
  requireSameShape(a, b);
  let sumSq = 0;
  for (let i = 0; i < a.length; i++) {
    const diff = b[i] - a[i];
    sumSq += diff * diff;
  }
  return Math.sqrt(sumSq);
}

/**
 * Return the original 2025 Marler Drift index.
 *
 * This historical descriptive index multiplies cosine distance by Euclidean
 * displacement. For unit-normalized embeddings those terms are algebraically
 * dependent, so v2 reports the components separately for interpretation.
 */
export function marlerDriftV1(a: Vector, b: Vector): number {
  return cosineDistance(a, b) * l2Distance(a, b);
}

export function driftComponents(a: Vector, b: Vector): DriftComponents {
  const cosDist = cosineDistance(a, b);
  const displacement = l2Distance(a, b);
  const normShift = norm(b) - norm(a);
  return {
    cosine_distance: cosDist,
    euclidean_displacement: displacement,
    norm_shift: normShift,
    marler_drift_v1: cosDist * displacement,
  };
}
